lib/Dialect/TT/IR/TTOpsTypes.cpp

```diff
@@ -513,8 +511,79 @@ MemorySpace LayoutAttr::getMemorySpace() const {
       .getValue();
 }
 
-mlir::AffineMap LayoutAttr::projectOnto(ArrayRef<int64_t> logicalTensorShape,
-                                        GridAttr grid) const {
+// Returns shape of the tensor after tilization is applied.
+llvm::SmallVector<int64_t>
+LayoutAttr::getTiledShape(llvm::ArrayRef<int64_t> tensorShape) const {
+  assert(isTiled() && "Expected a tiled layout");
+
+  auto tileLinearMap = getTileLinearMap();
+
+  mlir::SmallVector<mlir::AffineExpr> tensorShapeExprs(
+      llvm::map_range(tensorShape, [context = getContext()](std::int64_t e) {
+        return getAffineConstantExpr(e - 1, context);
+      }));
+
+  mlir::SmallVector<std::int64_t> tiledShape(tileLinearMap.getNumResults());
+  for (unsigned i = 0; i < tileLinearMap.getNumResults(); ++i) {
+    mlir::AffineExpr expr = tileLinearMap.getResult(i);
+    mlir::AffineExpr constantExpr = expr.replaceDims(tensorShapeExprs);
+    tiledShape[i] =
+        llvm::cast<mlir::AffineConstantExpr>(constantExpr).getValue() + 1;
+  }
+
+  return tiledShape;
+}
+
+// Returns an affine map that maps tensor original dimensions to the tile linear
+// space instead of scalar space.
+// Example:
+// scalar linear: (d0, d1, d2, d3) -> (d0 * 192 + d1 * 64 + d2, d3),
+// tile linear: (d0, d1, d2, d3) -> (d0 * 6 + d1 * 2 + d2 floordiv 32, d3
+// floordiv 32).
+mlir::AffineMap LayoutAttr::getTileLinearMap() const {
+  assert(isTiled() && "Expected a tiled layout");
+
+  mlir::AffineMap linear = getLinear();
+  mlir::SmallVector<AffineExpr> tileLinearExprs;
+
+  TileType tileType = mlir::cast<TileType>(getElementType());
+  int64_t tileH = tileType.getHeight();
+  int64_t tileW = tileType.getWidth();
+
+  mlir::AffineExpr d0 = getAffineDimExpr(0, getContext());
+  mlir::AffineExpr d1 = getAffineDimExpr(1, getContext());
+  tileLinearExprs.push_back(d0.floorDiv(tileH));
+  tileLinearExprs.push_back(d1.floorDiv(tileW));
```

Comment:
I think these should be `ceilDiv` right?

Reply:
`floorDiv` gives us 0-indexed mapping, right?

Reply:
But if this is for a shape calc we want to round up, i.e. tensor with shape 31x31 should round up to tile 32x32.

Reply:
This is part of `getTileLinearMap`, it doesn't directly affect shape calculation.

Reply:
I see, yeah ok since it goes through `evalShape` yeah I think it works.  I'm confusing things with the way I proposed in the top level comment.

Reply:
Even after I implemented it your way, floorDiv gave me the correct result. Example, consider tensor's dim is 64x128. If we got `(d0, d1)->(d0.ceilDiv(32), d1.ceilDiv(32))` and plug that in `evalShape`, 64 is decremented to 63 and ceilDiv gives 63/32=2, and then `evalShape` increments it to 3 which is incorrect.

Reply:
Yeah you're right, ceilDiv would incorrectly round up at perfectly aligned tile sizes!